services/ui/src/components/Problems/index.js

```diff
@@ -1,229 +1,368 @@
+
+  const [problemTerm, setProblemTerm] = useState('');
+  const [hasFilter, setHasFilter] = useState(false);
+  const [problemStats, setProblemStats] = useState([]);
+
+  const severities = getOptionsFromProblems(problems, 'severity');
+  const sources = getOptionsFromProblems(problems, 'source');
+  const services = getOptionsFromProblems(problems, 'service');
+
+
+  // Handlers
+  const handleProblemFilterChange = (event) => {
+    setHasFilter(false);
+
+    if (event.target.value !== null || event.target.value !== '') {
+      setHasFilter(true);
+    }
+    setProblemTerm(event.target.value);
+  };
+
+  const handleSourceChange = (source) => {
+    let values = source && source.map(s => s.value) || [];
+    setSource(values);
+  };
+
+  const handleSeverityChange = (severity) => {
+    let values = severity && severity.map(s => s.value) || [];
+    setSeverity(values);
+  };
+
+  const handleServiceChange = (service) => {
+    let values = service && service.map(s => s.value) || [];
+    setService(values);
+  };
+
+  // Options
+  const severityOptions = (severity) => {
+    return severity && severity.map(s => ({ value: s, label: s}));
+  };
+
+  const sourceOptions = (sources) => {
+    return sources && sources.map(s => ({ value: s, label: s}));
+  };
+
+  const serviceOptions = (services) => {
+    return services && services.map(s => ({ value: s, label: s}));
+  };
+
+  // Filters
+  const severityFilter = (item) => {
+    return Object.keys(item).some(key => {
+        if (item[key] !== null) {
+          return severitySelected.indexOf(item['severity'].toString()) > -1;
+        };
+    });
+  }
+
+  const sourceFilter = (item) => {
+    return Object.keys(item).some(key => {
+        if (item[key] !== null) {
+          return sourceSelected.indexOf(item['source'].toString()) > -1;
+        };
+    });
+  }
 
-    const [problemTerm, setProblemTerm] = useState('');
-    const [hasFilter, setHasFilter] = React.useState(false);
+  const serviceFilter = (item) => {
+    return Object.keys(item).some(key => {
+        if (item[key] !== null) {
+          return servicesSelected.indexOf(item['service'].toString()) > -1;
+        };
+    });
+  }
 
-    const handleProblemFilterChange = (event) => {
-        setHasFilter(false);
+  // Filter by select fields and text input field.
+  const filterResults = (item) => {
+    let filtered = true;
 
-        if (event.target.value !== null || event.target.value !== '') {
-            setHasFilter(true);
+    if (severitySelected.length > 0 && filtered) {
+      filtered = severityFilter(item);
+    }
+
+    if (sourceSelected.length > 0 && filtered) {
+      filtered = sourceFilter(item);
+    }
+
+    if (servicesSelected.length > 0 && filtered) {
+      filtered = serviceFilter(item);
+    }
+
+    // Filter text input
+    const lowercasedFilter = problemTerm.toLowerCase();
+    if (filtered && (problemTerm != null || problemTerm !== '')) {
```

Comment:
if you moved this into its own function (like the other three) then you could, I think, reduce this entire function to something like ...

shouldItemBeFiltered = (i) => isFilteredBySeverity(i) || isFilteredByService(i) || isFilteredbySource(i) || isFilteredByTextFilter(i)

I'm not suggesting that you do this kind of thing .. but I think we could get this a lot simpler

Reply:
moved into its own function and clean the whole function up